File: src/alphazero/mcts.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Dict, Iterable, Tuple

import numpy as np
import torch

from alphazero.action_helper import EurobotActionHelper
from alphazero.network import AlphaZeroNetwork


@dataclass
class MCTSConfig:
    num_simulations: int = 128
    c_puct: float = 1.5
    dirichlet_alpha: float = 0.3
    dirichlet_epsilon: float = 0.25


class TreeNode:
    __slots__ = ("prior", "visit_count", "value_sum", "children", "is_expanded")

    def __init__(self, prior: float = 0.0) -> None:
        self.prior: float = float(prior)
        self.visit_count: int = 0
        self.value_sum: float = 0.0
        self.children: Dict[int, "TreeNode"] = {}
        self.is_expanded: bool = False

    @property
    def value(self) -> float:
        if self.visit_count == 0:
            return 0.0
        return self.value_sum / self.visit_count


class MCTS:
    def __init__(
        self,
        network: AlphaZeroNetwork,
        action_helper: EurobotActionHelper,
        config: MCTSConfig,
        device: torch.device,
    ) -> None:
        self.network = network
        self.helper = action_helper
        self.cfg = config
        self.device = device

    def search(
        self,
        env,
        root_obs: np.ndarray,
        root_state,
        add_dirichlet_noise: bool = True,
    ) -> TreeNode:
        env.set_state(root_state)
        root = TreeNode(prior=1.0)
        self._expand(root, root_obs, env)
        if add_dirichlet_noise and root.children:
            self._add_dirichlet_noise(root)

        for _ in range(max(1, self.cfg.num_simulations)):
            env.set_state(root_state)
            self._simulate(env, root_obs, root)
        return root

    def _simulate(self, env, root_obs: np.ndarray, root: TreeNode) -> None:
        '''
        Traverse down the tree using (P)UCT and expand the first unexpanded node.
        Backprop the value back to root.
        '''
        node = root
        obs = np.array(root_obs, copy=True)
        path = [node]

        while True:
            if not node.is_expanded:
                value = self._expand(node, obs, env)
                break
            action_idx, child = self._select_child(node)
            action = self.helper.decode(action_idx)
            obs, _, terminated, truncated, _ = env.step(np.asarray(action, dtype=np.int64)) # da li mi treba ovo??
            path.append(child)
            node = child
            done = bool(terminated) or bool(truncated)
            if done:
                value = self._terminal_value(env)
                break

        self._backprop(path, value)

    def _expand(self, node: TreeNode, obs: np.ndarray, env) -> float:
        '''
        Expands current node with all legal actions and assigns prior to all of them
        using policy network.
        '''
        mask = self.helper.legal_action_mask(obs)
        if not np.any(mask):
            print("Ne postoji validna akcija za ovo stanje!")
            node.children = {}
            node.is_expanded = True
            return self._terminal_value(env)

        obs_tensor = torch.as_tensor(obs, dtype=torch.float32, device=self.device).unsqueeze(0)
        with torch.no_grad():
            (lv, ln, lc, lq), value = self.network(obs_tensor)  # each [1, dim]
        lv = lv.squeeze(0).detach().cpu()
        ln = ln.squeeze(0).detach().cpu()
        lc = lc.squeeze(0).detach().cpu()
        lq = lq.squeeze(0).detach().cpu()

        # fast composition: joint_logits[a] = lv[v[a]] + ln[n[a]] + lc[c[a]] + lq[q[a]]
        verb_idx  = torch.as_tensor(self.helper.act_verb,  dtype=torch.long)
        node_idx  = torch.as_tensor(self.helper.act_node,  dtype=torch.long)
        color_idx = torch.as_tensor(self.helper.act_color, dtype=torch.long)
        qty_idx   = torch.as_tensor(self.helper.act_qty,   dtype=torch.long)

        joint = lv[verb_idx] + ln[node_idx] + lc[color_idx] + lq[qty_idx]  # [A] torch
        joint_np = joint.numpy()

        # masked softmax over legal actions 
        priors = self._masked_softmax(joint_np, mask)
        node.children = {idx: TreeNode(prior=float(priors[idx])) for idx, valid in enumerate(mask) if valid}
        node.is_expanded = True
        return float(value.item())

    def _select_child(self, node: TreeNode) -> Tuple[int, TreeNode]:
        '''
        finds the child of the node with the best (P)UCT score
        '''
        best_score = -float("inf")
        best_action = None
        best_child = None
        sqrt_total = math.sqrt(node.visit_count + 1.0)
        for action, child in node.children.items():
            q = child.value
            u = self.cfg.c_puct * child.prior * sqrt_total / (1.0 + child.visit_count)
            score = q + u
            if score > best_score:
                best_score = score
                best_action = action
                best_child = child
        if best_child is None or best_action is None:
            raise RuntimeError("MCTS: failed to select a child node. Check action masking.")
        return best_action, best_child

    def _backprop(self, path: Iterable[TreeNode], value: float) -> None:
        for node in reversed(list(path)):
            node.visit_count += 1
            node.value_sum += value

    def _add_dirichlet_noise(self, node: TreeNode) -> None:
        actions = list(node.children.keys())
        if not actions:
            return
        alpha = self.cfg.dirichlet_alpha
        epsilon = self.cfg.dirichlet_epsilon
        noise = np.random.dirichlet([alpha] * len(actions))
        for action, n in zip(actions, noise):
            child = node.children[action]
            child.prior = (1 - epsilon) * child.prior + epsilon * float(n)

    def _masked_softmax(self, logits: np.ndarray, mask: np.ndarray) -> np.ndarray:
        masked_logits = np.where(mask, logits, -np.inf)
        finite = masked_logits[np.isfinite(masked_logits)]
        if finite.size == 0:
            probs = mask.astype(np.float32)
            total = probs.sum()
            return probs / max(total, 1.0)
        max_logit = np.max(finite)
        exp_logits = np.exp(masked_logits - max_logit)
        exp_logits = np.where(mask, exp_logits, 0.0)
        total = exp_logits.sum()
        if total <= 0.0:
            probs = mask.astype(np.float32)
            total = probs.sum()
            return probs / max(total, 1.0)
        return exp_logits / total

    def _terminal_value(self, env) -> float:
        blue_score, _ = env.world.final_scores_norm()
        return blue_score

File: src/alphazero/mcts.cpp
#include "alphazero/mcts.hpp"

#include <ATen/core/Reduction.h>
#include <cmath>
#include <torch/torch.h>

#include <stdexcept>
#include <utility>

namespace alphazero {

MCTS::MCTS(MCTSConfig config, PolicyNetwork &network, ActionHelper &action_helper)
    : config_(std::move(config)), policy_network_(network), action_helper_(action_helper) {
  if (config_.num_simulations <= 0) {
    throw std::invalid_argument("MCTS: simulations must be positive");
  }
}

/*
 * Searches with mcts to find the best possible action for current root_state.
 * This can be seen as main MCTS function.
 */
TreeNode MCTS::search(eurobot::EurobotWorld &world) {

    const eurobot::EurobotState root_state = world.get_state();
    auto root = std::make_shared<TreeNode>(1.0f);

    expand(world, root);
    add_dirichlet_noise(root);

    for(int i = 0; i < config_.num_simulations; i++){
        world.set_state(root_state);
        simulate(world, root);
    }
    world.set_state(root_state);

    return *root;
}

/*
 * Expandes the node with all possible legal actions in current state of the world
 * and assigns prior to each child using policy network
 */
double MCTS::expand(eurobot::EurobotWorld &world, std::shared_ptr<TreeNode>& node){
    // run obs through policy network
    c10::InferenceMode guard;
    auto [logits, value] = policy_network_->forward(world.obs());
    logits = logits.squeeze(0);
    TORCH_CHECK(logits.dim() == 1, "policy logits must be 1D [A]");
    TORCH_CHECK(logits.dtype() == torch::kFloat32, "policy logits must be float32");
    TORCH_CHECK(logits.size(0) == (long)action_helper_.actions().size(),
                "policy logits size != action space size");

    // TODO: collect multi-head logits into one [#action_size] tensor

    // mask it with legal_action_mask
    auto logit_mask = action_helper_.logit_mask(world);
    auto masked_logits = logits + logit_mask;
    masked_logits = torch::where(torch::isfinite(masked_logits),
                                 masked_logits,
                                 torch::full_like(masked_logits, -1e9f));
    auto priors = torch::softmax(masked_logits, -1).contiguous();

    const int64_t A = priors.size(0);
    auto acc = priors.data_ptr<float>(); ;

    node->children.clear();
    node->children.reserve(A);
    for (int64_t a = 0; a < A; ++a) {
    float p = acc[a];
        if (p > 0.0f && std::isfinite(p)) {
            node->children.emplace_back((int)a, std::make_shared<TreeNode>(p));
        }
    }
    node->is_expanded = true;
    return value.item<double>();
}

/*
 * Selects the action of the expanded node using (P)UCT.
 */
std::pair<int, std::shared_ptr<TreeNode>> MCTS::select_child(const std::shared_ptr<TreeNode>& node){
    double best_score = -std::numeric_limits<double>::infinity();
    int best_action = -1;
    std::shared_ptr<TreeNode> best_child;
    const double sqrt_total = std::sqrt(node->vis_count + 1.0);

    for (const auto& kv : node->children) {
        const int a = kv.first;
        const auto& child = kv.second;

        double q = child->value();
        double u = config_.cpuct * child->prior * sqrt_total / (1.0 + child->vis_count);
        double score = q + u;
        if(score > best_score){
            best_score = score;
            best_action = a;
            best_child = child;
        }
    }
    return {best_action, best_child};
}

/*
 * Traverse the tree from the root until first unexpanded node, or
 * until end of the episode. At the end, backup the traversed path up the tree.
 */
void MCTS::simulate(eurobot::EurobotWorld &world, const std::shared_ptr<TreeNode>& root){
    std::vector<std::shared_ptr<TreeNode>> path;
    auto node = root;
    path.push_back(node);

    double value = 0.0;

    while(true){
        if(!node->is_expanded){
            value = expand(world, node);
            break;
        }
        auto [action_idx, child] = select_child(node);
        const auto &action = action_helper_.action_by_index(action_idx);
        const bool done = world.step_blue(action);

        path.push_back(child);
        node = child;

        if(done){
            value = world.final_scores_norm().first;
            break;
        }
    }

    // backprop
    for (int i = (int)path.size() - 1; i >= 0; --i) {
        path[i]->vis_count += 1;
        path[i]->value_sum += value;
    }
}

/*
 * Adds dirichlet noise to the root children priors
 */
void MCTS::add_dirichlet_noise(std::shared_ptr<TreeNode>& node) {
  const float eps   = config_.dirichlet_epsilon;
  const float alpha = config_.dirichlet_alpha;
  if (node->children.empty() || eps <= 0.f || alpha <= 0.f) return;

  // sample Dirichlet(alpha) via Gamma(alpha,1)
  std::vector<float> noise(node->children.size());
  std::gamma_distribution<float> gamma(alpha, 1.0f);
  float sum = 0.f;
  static thread_local std::mt19937 rng{std::random_device{}()};

  for (auto &z : noise) { z = std::max(1e-12f, gamma(rng)); sum += z; }
  const float inv_sum = 1.0f / sum;
  for (auto &z : noise) z *= inv_sum;

  // mix into child priors
  for (size_t i = 0; i < node->children.size(); ++i) {
    auto& child = node->children[i].second;
    child->prior = (1.0f - eps) * child->prior + eps * noise[i];
  }
}


} // namespace alphazero
